doc/contributing/erm-guidelines.md

```diff
@@ -0,0 +1,403 @@
+   has not been explicitly closed when the disposal method is called, the
+   object should be disposed as if an exception had occurred. For instance,
+   if the object API exposes both a `close()` method and an `abort()` method,
+   the disposal method should call `abort()` if the object is not already
+   closed. If there is no difference in disposing in success or exception
+   contexts, then separate disposal methods are unnecessary.
+3. It is recommended to avoid throwing errors within disposers.
+   If a disposer throws an exception while there is another pending
+   exception, then both exceptions will be wrapped in a `SupressedError`
+   that masks both. This makes it difficult to understand the context
+   in which the exceptions were thrown.
+4. Disposable objects should expose explicit disposal methods in addition
+   to the `Symbol.dispose` and `Symbol.asyncDispose` methods. This allows
+   user code to explicitly dispose of the object without using the `using`
+   or `await using` statements. For example, a disposable object might
+   expose a `close()` method that can be called to dispose of the object.
+   The `Symbol.dispose` and `Symbol.asyncDispose` methods should delegate to
+   these explicit disposal methods.
+5. Because it is safest to assume that the disposal method will be called
+   in an exception context, it is generally recommended to prefer use of
+   `Symbol.dispose` over `Symbol.asyncDispose` when possible. Asynchronous
+   disposal can lead delaying the handling of exceptions and can make it
+   difficult to reason about the state of the object while the disposal is
+   in progress. Disposal in an exception context is preferably synchronous
+   and immediate. That said, for some types of objects async disposal is not
+   avoidable.
+6. Asynchronous disposers, by definition, are able to yield to other tasks
+   while waiting for their disposal task(s) to complete. This means that, as a
+   minimum, a `Symbol.asyncDispose` method must be an `async` function, and
+   must `await` at least one asynchronous disposal task. If either of these
+   criteria is not met, then the disposer is actually a synchronous disposer in
+   disguise, and will block the execution thread until it returns; such a
+   disposer should instead be declared using `Symbol.dispose`.
+7. Avoid, as much as possible, using both `Symbol.dispose` and `Symbl.asyncDispose`
+   in the same object. This can make it difficult to reason about which method
+   will be called in a given context and could lead to unexpected behavior or
+   subtle bugs. This is not a firm rule, however. Sometimes it may make sense
+   to define both but likely not.
+
+### Example Disposable Objects
+
+A disposable object can be quite simple:
+
+```js
+class MyResource {
+  dispose() {
+    console.log('Resource disposed');
+  }
+
+  [Symbol.dispose]() {
+    this.dispose();
+  }
+}
+
+{ using myDisposable = new MyResource(); }
+```
+
+Or even fully anonymous objects:
```

Comment:
I'd disagree with recommending returning an anonymous objects in an API design guideline. We should recommend structured API to improve comprehensiveness and discoverablity.

Reply:
@legendecas The problem is that lots of APIs need a bespoke disposable which isn't used for anything else, and adding a new class for every such API ends up littering lots of little classes all over the place. See e.g. https://github.com/nodejs/node/pull/58516 for such an API. I think anonymous objects are better than having a separate class for each such API.

Reply:
I think the question is more like how user anticipates with a disposable API. If an API returns an object, it should be easy to know what users can do with the object. 

https://github.com/nodejs/node/pull/58516 is an example that there are more properties than the documented `path` on the returned anonymous object, like `remove()`. It is not ideal to just say "the returned object has property A, B, and C". We should document every public API and its behavior clearly if possible.

Regarding "adding a new class for every such API ends up littering lots of little classes all over the place", IMO this is a question on if it is worthwhile to add a new API to add disposable support. If an API already returns an object, there is no need to add a new disposable class for it.

Reply:
I'm not suggesting that we would leave things undocumented. I'm saying that we don't always need formal classes in the actual implementation. For some return values we can safely rely on documentation-only and leave the actual implementation as anonymous objects. We even have existing precedence for this in the current documentation. See, for instance, all of the "Class" documentations for Web Crypto operations here: https://nodejs.org/docs/latest/api/webcrypto.html#algorithm-parameters ... the things like `AesDerivedKeyParams` is documented as a class but, in reality, there is no actual class named `AesDerivedKeyParams` in the source.

Reply:
I've expanded the document to include coverage of documentation of anonymous dispoables.